flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java

```diff
@@ -735,6 +739,56 @@ public StreamGraph generate() {
 		return Collections.singleton(transform.getId());
 	}
 
+	private <OUT> Collection<Integer> transformMultipleInputTransform(MultipleInputTransformation<OUT> transform) {
+		List<Collection<Integer>> allInputIds = new ArrayList<>();
+
+		for (Transformation<?> input : transform.getInputs()) {
+			allInputIds.add(transform(input));
+		}
+
+		// the recursive call might have already transformed this
+		if (alreadyTransformed.containsKey(transform)) {
+			return alreadyTransformed.get(transform);
+		}
+
```

Comment:
I'm not sure. I was mostly copying the logic from `transformTwoInputTransform` (and also the same pattern is in `transformOneInputTransform` as well).

However it looks like this might be because of cyclic dependencies. 

Do you think this is a bug in all of those places?

Reply:
Or maybe because of mutability of `Transformation`?
In either case, I wouldn't spend time on it because the only impact I see is this confusion.